js/state/syncUtil.js

```diff
@@ -20,3 +34,124 @@ module.exports.setObjectId = (item) => {
+  return Math.floor(Date.now() / 1000)
+}
+
+/**
+ * Checks whether an object is syncable as a record of the given type
+ * @param {string} type
+ * @param {Immutable.Map} item
+ * @returns {boolean}
+ */
+module.exports.isSyncable = (type, item) => {
+  if (type === 'bookmark' && item.get('tags')) {
+    return (item.get('tags').includes('bookmark') ||
+      item.get('tags').includes('bookmark-folder'))
+  } else if (type === 'siteSetting') {
+    for (let field in siteSettingDefaults) {
+      if (item.has(field)) {
+        return true
+      }
+    }
+  }
+  return false
+}
+
+/**
+ * Sets a new object ID for an existing object in appState
+ * @param {Array.<string>} objectPath - Path to get to the object from appState root,
+ *   for use with Immutable.setIn
+ * @returns {Array.<number>}
+ */
+module.exports.newObjectId = (objectPath) => {
+  const crypto = require('crypto')
+  const appActions = require('../actions/appActions')
+  const objectId = new Immutable.List(crypto.randomBytes(16))
+  appActions.setObjectId(objectId, objectPath)
+  return objectId.toJS()
+}
+
+/**
+ * Converts a site object into input for sendSyncRecords
+ * @param {Object} site
+ * @param {number=} siteIndex
+ * @returns {{name: string, value: object, objectId: Array.<number>}}
+ */
+module.exports.createSiteData = (site, siteIndex) => {
+  const siteData = {
+    location: '',
+    title: '',
+    customTitle: '',
+    lastAccessedTime: 0,
+    creationTime: 0
+  }
+  for (let field in site) {
+    if (field in siteData) {
+      siteData[field] = site[field]
+    }
+  }
+  if (module.exports.isSyncable('bookmark', Immutable.fromJS(site))) {
+    if (!site.objectId && typeof siteIndex !== 'number') {
+      throw new Error('Missing bookmark objectId.')
+    }
+    return {
+      name: 'bookmark',
+      objectId: site.objectId || module.exports.newObjectId(['sites', siteIndex]),
+      value: {
+        site: siteData,
+        isFolder: site.tags.includes('bookmark-folder'),
+        folderId: site.folderId || 0,
+        parentFolderId: site.parentFolderId || 0
+      }
+    }
+  } else if (!site.tags || !site.tags.length) {
+    if (!site.objectId && typeof siteIndex !== 'number') {
+      throw new Error('Missing historySite objectId.')
+    }
+    return {
+      name: 'historySite',
+      objectId: site.objectId || module.exports.newObjectId(['sites', siteIndex]),
+      value: siteData
+    }
+  }
+}
+
+/**
+ * Converts a site settings object into input for sendSyncRecords
+ * @param {string} hostPattern
+ * @param {Object} setting
+ * @returns {{name: string, value: object, objectId: Array.<number>}}
+ */
+module.exports.createSiteSettingsData = (hostPattern, setting) => {
+  const adControlEnum = {
+    showBraveAds: 0,
+    blockAds: 1,
+    allowAdsAndTracking: 2
+  }
+  const cookieControlEnum = {
+    block3rdPartyCookie: 0,
+    allowAllCookies: 1
+  }
+  const value = Object.assign({hostPattern}, siteSettingDefaults)
```

Comment:
i guess it should actually be `Object.assign({}, siteSettingDefaults, {hostPattern})`